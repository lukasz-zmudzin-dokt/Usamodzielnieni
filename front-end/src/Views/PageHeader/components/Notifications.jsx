import React, { useState, useContext, useEffect } from "react";
import { Nav, Dropdown, Col } from "react-bootstrap";
import NotificationItemContainer from "./NotificationItemContainer";
import NotificationItem from "./NotificationItem";
import NotificationToggle from "./NotificationToggle";
import { NotificationsContext } from "context";

const Notifications = ({ location, token, ...rest }) => {
  const [show, setShow] = useState(false);
  const notificationsContext = useContext(NotificationsContext);
  const { notifications, count, error } = notificationsContext;

  useEffect(() => {
    if (notifications && count) {
      const toRemove = notifications.filter(
        (notification) => notification.path === location.pathname
      );
      toRemove.forEach((notification) => {
        notificationsContext.deleteNotification(notification.id);
      });
    }
  }, [notificationsContext, count, notifications, location.pathname]);

  const clearNotifications = async () => {
    setShow("prevent");
    await notificationsContext.deleteNotifications();
  };

  const removeNotification = async (id) => {
    await notificationsContext.deleteNotification(id);
  };

  const onToggle = async (isOpen) => {
    setShow((prev) => (prev !== "prevent" ? !prev : true));
    if (isOpen && count) {
      await notificationsContext.markAsRead();
    }
  };

  const loadMoreNotifications = async () => {
    setShow("prevent");
    await notificationsContext.loadMoreNotifications();
  };

  return (
    <Dropdown
      as={Nav.Item}
      alignRight
      show={show}
      onToggle={onToggle}
      {...rest}
    >
      <Dropdown.Toggle
        as={NotificationToggle}
        count={notifications === null || error ? 0 : count}
      />
      <Dropdown.Menu data-testid="dropdownMenu">
        {notifications === null ? (
          <Dropdown.Item disabled="true">Ładowanie...</Dropdown.Item>
        ) : (
          <>
            {!notifications.length ? (
              <Dropdown.Item as={Col} disabled>
                Brak powiadomień
              </Dropdown.Item>
            ) : (
              <div className="notifications-container">
                {notifications.map((notification) => (
                  <NotificationItemContainer
                    key={notification.id}
                    data-testid={`dropdownItem${notification.id}`}
                    onClick={removeNotification}
                    notification={notification}
                  >
                    <Dropdown.Item
                      as={NotificationItem}
                      notification={notification}
                    />
                  </NotificationItemContainer>
                ))}
                {notificationsContext.next !== undefined && (
                  <Dropdown.Item
                    key="loadButton"
                    onClick={loadMoreNotifications}
                    disabled={notificationsContext.next === null}
                  >
                    Załaduj kolejne
                  </Dropdown.Item>
                )}
              </div>
            )}
            <Dropdown.Divider />
            <Dropdown.Item
              key="clearButton"
              onClick={clearNotifications}
              disabled={!notifications.length}
            >
              Wyczyść
            </Dropdown.Item>
          </>
        )}
      </Dropdown.Menu>
    </Dropdown>
  );
};

export default Notifications;
