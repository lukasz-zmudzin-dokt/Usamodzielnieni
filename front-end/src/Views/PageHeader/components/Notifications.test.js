import React from "react";
import { render, fireEvent } from "@testing-library/react";
import Notifications from "./Notifications";
import { MemoryRouter } from 'react-router-dom';

describe('Notifications', () => {
    it('should render without crashing', () => {

        const { container } = render(
            <MemoryRouter>
                <Notifications />
            </MemoryRouter>
        );

        expect(container).toMatchSnapshot();
    });

    it('should open notifications list when toggle is clicked', async () => {

        const { container, getByText } = render(
            <MemoryRouter>
                <Notifications />
            </MemoryRouter>
        );
        fireEvent.click(getByText('Powiadomienia', { exact: false }));

        await expect(container).toMatchSnapshot();
    });
});