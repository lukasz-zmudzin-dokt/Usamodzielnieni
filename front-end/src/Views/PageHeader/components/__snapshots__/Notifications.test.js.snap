// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Notifications should open notifications list when toggle is clicked and api returns list with 2 objects 1`] = `
<div
  aria-labelledby=""
  class="dropdown-menu show"
  data-testid="dropdownMenu"
  style="position: absolute; top: 0px; left: 0px; transform: translate(0px, 0px);"
  x-placement="bottom-start"
>
  <div
    class="notifications-container"
  >
    <div
      class="notificationItemContainer btn-group"
      data-testid="dropdownItem1"
      role="group"
    >
      <a
        class="dropdown-item"
        href="/cvEditor"
      >
        <div>
          Rozpatrzono CV
        </div>
        <small>
          3/14/2020
        </small>
      </a>
      <button
        class="btn btn-light"
        type="button"
      >
        X
      </button>
    </div>
    <div
      class="notificationItemContainer btn-group"
      data-testid="dropdownItem2"
      role="group"
    >
      <a
        class="dropdown-item"
        href="/"
      >
        <div>
          Dostępne nowe oferty pracy
        </div>
        <small>
          3/14/2020
        </small>
      </a>
      <button
        class="btn btn-light"
        type="button"
      >
        X
      </button>
    </div>
  </div>
  <div
    class="dropdown-divider"
    role="separator"
  />
  <a
    class="dropdown-item"
    href="#"
    role="button"
  >
    Wyczyść
  </a>
</div>
`;

exports[`Notifications should render without crashing 1`] = `
<div>
  <div
    class="dropdown nav-item"
  >
    <button
      aria-expanded="false"
      aria-haspopup="true"
      class="dropdown-toggle btn btn-light"
      type="button"
    >
      Powiadomienia 
    </button>
  </div>
</div>
`;
